pkg/schemacompat/schemacompat.go

```diff
@@ -0,0 +1,374 @@
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Not, new.Not, "not", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "allOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "anytOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "oneOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Enum, new.Enum, "enum", typeName)...)
+	if !(new.Maximum == existing.Maximum && new.Minimum == existing.Minimum && new.ExclusiveMaximum == existing.ExclusiveMaximum && new.ExclusiveMinimum == existing.ExclusiveMinimum) {
+		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Maximum, new.Maximum, "maximum", typeName)...)
+		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Minimum, new.Minimum, "minimum", typeName)...)
+	}
+	if new.MultipleOf != existing.MultipleOf {
+		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.MultipleOf, new.MultipleOf, "multipleOf", typeName)...)
+	}
+	return
+}
+
+func lcdForStructural(fldPath *field.Path, existing, new *schema.Structural, lcd *schema.Structural, narrowExisting bool) (errorList field.ErrorList) {
+	if lcd == nil && narrowExisting {
+		return field.ErrorList{field.InternalError(fldPath, errors.New("lcd argument should be passed when narrowExisting is true"))}
+	}
+	if new == nil {
+		return field.ErrorList{field.Invalid(fldPath, nil, "new schema doesn't allow anything")}
+	}
+	if existing.XPreserveUnknownFields != new.XPreserveUnknownFields {
+		errorList = append(errorList, field.Invalid(fldPath.Child("x-preserve-unknown-fields"), new.XPreserveUnknownFields, "x-preserve-unknown-fields value has been changed in an incompatible way"))
+	}
+
+	switch existing.Type {
+	case "number":
+		return lcdForNumber(fldPath, existing, new, lcd, narrowExisting)
+	case "integer":
+		return lcdForInteger(fldPath, existing, new, lcd, narrowExisting)
+	case "string":
+		return lcdForString(fldPath, existing, new, lcd, narrowExisting)
+	case "boolean":
+		return lcdForBoolean(fldPath, existing, new, lcd, narrowExisting)
+	case "array":
+		return lcdForArray(fldPath, existing, new, lcd, narrowExisting)
+	case "object":
+		return lcdForObject(fldPath, existing, new, lcd, narrowExisting)
+	case "":
+		if existing.XIntOrString {
+			return lcdForIntOrString(fldPath, existing, new, lcd, narrowExisting)
+		} else if existing.XPreserveUnknownFields {
+			return lcdForPreserveUnknownFields(fldPath, existing, new, lcd, narrowExisting)
+		}
+	}
+	return field.ErrorList{field.Invalid(field.NewPath(fldPath.String(), "type"), existing.Type, "Invalid type")}
+}
+
+func lcdForIntegerValidation(fldPath *field.Path, existing, new *schema.ValueValidation, lcd *schema.ValueValidation, narrowExisting bool) (errorList field.ErrorList) {
```

Comment:
These should also just return `error` and you won't have to deal with appending to an `ErrorList`.

Reply:
I don't see why here. `checkUnsupportedValidationForNumerics` may return several incompatibility errors (for min/max interval constraints, and for the `multipleOf` constraint.